# Week3 Pilot Task

In [1]:
import pandas as pd

shots_df=pd.read_csv('processing_shots_df.csv')

shots_df.head()

id  minute      X      Y        xG           player  player_id  season  \
0  585157      28  0.900  0.344  0.074058     Bryan Mbeumo       6552    2024   
1  585158      30  0.956  0.610  0.373126  Kristoffer Ajer       9677    2024   
2  585159      41  0.920  0.669  0.078655     Kevin Schade       9156    2024   
3  585160      46  0.850  0.534  0.011295    Ethan Pinnock       9678    2024   
4  585162      51  0.914  0.652  0.039136    Vitaly Janelt       9680    2024   

   match_id     h_team  ... assist_flag  home_away  situation_FromCorner  \
0     26609  Brentford  ...           1          1                     0   
1     26609  Brentford  ...           1          1                     0   
2     26609  Brentford  ...           1          1                     0   
3     26609  Brentford  ...           1          1                     1   
4     26609  Brentford  ...           1          1                     1   

  situation_OpenPlay situation_Penalty situation_SetPiece  shotType_LeftFoot  \
0                  1                 0                  0                  1   
1                  1                 0                  0                  0   
2                  1                 0                  0                  0   
3                  0                 0                  0                  0   
4                  0                 0                  0                  1   

   shotType_OtherBodyPart  shotType_RightFoot  lastAction_encoded  
0                       0                   0            0.099097  
1                       0                   1            0.272727  
2                       0                   1            0.105516  
3                       0                   0            0.042131  
4                       0                   0            0.130534  

[5 rows x 29 columns]

In [2]:
shots_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9878 entries, 0 to 9877
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      9878 non-null   int64  
 1   minute                  9878 non-null   int64  
 2   X                       9878 non-null   float64
 3   Y                       9878 non-null   float64
 4   xG                      9878 non-null   float64
 5   player                  9878 non-null   object 
 6   player_id               9878 non-null   int64  
 7   season                  9878 non-null   int64  
 8   match_id                9878 non-null   int64  
 9   h_team                  9878 non-null   object 
 10  a_team                  9878 non-null   object 
 11  h_goals                 9878 non-null   int64  
 12  a_goals                 9878 non-null   int64  
 13  date                    9878 non-null   object 
 14  player_assisted         9878 non-null   

In [4]:
# 모델 학습에 사용할 변수
features = [
    "X", "Y",
    "shot_distance", "shot_angle",
    "assist_flag", "home_away",
    "situation_FromCorner", "situation_OpenPlay",
    "situation_Penalty", "situation_SetPiece",
    "shotType_LeftFoot", "shotType_RightFoot", "shotType_OtherBodyPart",
    "lastAction_encoded"
]
original_xG=shots_df['xG'] # 실제 xG값
target = "is_goal"

In [5]:
shots_df[features].describe()

X            Y  shot_distance   shot_angle  assist_flag  \
count  9878.000000  9878.000000    9878.000000  9878.000000  9878.000000   
mean      0.858118     0.502775       0.184989     0.577299     0.746811   
std       0.086710     0.120866       0.089683     0.376306     0.434860   
min       0.006000     0.056000       0.008000     0.002676     0.000000   
25%       0.808000     0.415000       0.119038     0.320075     0.000000   
50%       0.872000     0.503000       0.182989     0.451566     1.000000   
75%       0.915000     0.588000       0.239366     0.734803     1.000000   
max       0.994000     0.995000       0.994652     2.847435     1.000000   

         home_away  situation_FromCorner  situation_OpenPlay  \
count  9878.000000           9878.000000         9878.000000   
mean      0.529459              0.163900            0.740838   
std       0.499157              0.370203            0.438197   
min       0.000000              0.000000            0.000000   
25%       0.000000              0.000000            0.000000   
50%       1.000000              0.000000            1.000000   
75%       1.000000              0.000000            1.000000   
max       1.000000              1.000000            1.000000   

       situation_Penalty  situation_SetPiece  shotType_LeftFoot  \
count        9878.000000         9878.000000        9878.000000   
mean            0.008403            0.059627           0.332861   
std             0.091284            0.236807           0.471261   
min             0.000000            0.000000           0.000000   
25%             0.000000            0.000000           0.000000   
50%             0.000000            0.000000           0.000000   
75%             0.000000            0.000000           1.000000   
max             1.000000            1.000000           1.000000   

       shotType_RightFoot  shotType_OtherBodyPart  lastAction_encoded  
count         9878.000000             9878.000000         9878.000000  
mean             0.487852                0.004252            0.109435  
std              0.499878                0.065071            0.048092  
min              0.000000                0.000000            0.000000  
25%              0.000000                0.000000            0.099097  
50%              0.000000                0.000000            0.099097  
75%              1.000000                0.000000            0.130534  
max              1.000000                1.000000            0.285714

In [6]:
from sklearn.model_selection import train_test_split

X = shots_df[features]
y = shots_df[target]

X_train, X_test, y_train, y_test, xG_train, xG_test = train_test_split(
    X, y, original_xG, test_size=0.2, random_state=42
) # xG값을 나중에 비교하기 위해 같은 인덱스로 분리해둠

In [7]:
print("학습 데이터 형태:", X_train.shape)
print("테스트 데이터 형태:", X_test.shape)

학습 데이터 형태: (7902, 14)
테스트 데이터 형태: (1976, 14)


In [8]:
import pandas as pd
import numpy as np

# 모델 라이브러리 임포트
import lightgbm as lgb
import xgboost as xgb
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, r2_score


X_train, X_test, y_train, y_test, xG_train, xG_test = train_test_split(
    X, y, original_xG, test_size=0.2, random_state=42
)


# 여러 모델 정의 
models = {
    "LightGBM": lgb.LGBMClassifier(random_state=42),
    "XGBoost": xgb.XGBClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "Logistic Regression": LogisticRegression(random_state=42),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "KNN": KNeighborsClassifier(),
}


# 모델 학습 및 성능 비교
# 각 모델의 성능을 저장할 딕셔너리
results = {}

for name, model in models.items():
    print(f"--- {name} 모델 학습 중... ---")
    
    # 모델 학습
    model.fit(X_train, y_train)
    
    # xG 예측 (골이 될 확률)
    predicted_xG = model.predict_proba(X_test)[:, 1]
    
    # 성능 평가 (실제 xG 값과 비교)
    mae = mean_absolute_error(xG_test, predicted_xG)
    r2 = r2_score(xG_test, predicted_xG)
    
    # 결과 저장
    results[name] = {'MAE': mae, 'R2': r2}
    print(f"{name} 모델 평가 완료: MAE={mae:.4f}, R2={r2:.4f}\n")


# 최종 결과 비교
print("--- 최종 모델 성능 비교 ---")
results_df = pd.DataFrame(results).T # .T는 행과 열을 바꿈 (Transpose)
print(results_df.sort_values(by='R2', ascending=False)) # R2가 높은 순으로 정렬하여 출력

--- LightGBM 모델 학습 중... ---
[LightGBM] [Info] Number of positive: 856, number of negative: 7046
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001117 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1049
[LightGBM] [Info] Number of data points in the train set: 7902, number of used features: 14
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.108327 -> initscore=-2.107945
[LightGBM] [Info] Start training from score -2.107945
LightGBM 모델 평가 완료: MAE=0.0703, R2=0.5287

--- XGBoost 모델 학습 중... ---
XGBoost 모델 평가 완료: MAE=0.0803, R2=0.3717

--- Random Forest 모델 학습 중... ---
Random Forest 모델 평가 완료: MAE=0.0826, R2=0.4200

--- Logistic Regression 모델 학습 중... ---
Logistic Regression 모델 평가 완료: MAE=0.0574, R2=0.6475

--- Decision Tree 모델 학습 중... ---
Decision Tree 모델 평가 완료: MAE=0.1772, R2=-2.2673

--- KNN 모델 학습 중... ---
KNN 모델 평가 완료: MAE=0.1124, R2=0.1

결정계수를 확인한 이유
- 분류모델은 도구일 뿐 실제로는 0과 1사이의 회귀값

MAE가 직접적인 오차의 크기, 결정계수가 설명력을 보여줌

트리기반의 모델보다 logistic 회귀가 더 잘 예측하는 것을 볼 수 있음

로지스틱회귀 확인한 이유
- 로지스틱 회귀는 0과 1 사이의 시그모이드 값을 가져 적합하다고 판단
- 각 변수와 결과 사이가 선형이라 가정하는 한계가 있음
- 비선형 관계를 학습하기 위해서는 직접 피쳐 엔지니어링을 해줘야함

linearegression은 분류에 적합하기에 제외

나이브베이즈는 독립성 가정이라 제외

In [9]:
model = LogisticRegression()
model.fit(X_train, y_train)


# 회귀 계수 확인 (가장 기본적인 방법)
print("--- 기본 출력 (Numpy 배열) ---")
print(model.coef_)
print("\n" + "="*30 + "\n")


# 피처 이름과 함께 보기 좋게 출력 (추천)
# pandas Series를 활용하여 피처 이름과 계수를 짝지어줍니다.
coefficients = pd.Series(model.coef_[0], index=X_train.columns)

print("--- 피처별 회귀 계수 ---")
print(coefficients.sort_values(ascending=False)) # 영향력이 큰 순으로 정렬

# 절편(intercept) 확인
print("\n--- 절편 (Intercept) ---")
print(model.intercept_[0])

--- 기본 출력 (Numpy 배열) ---
[[ 3.05219982 -0.25790758 -2.21813263  1.76316398  0.18714555 -0.09964128
  -0.47672952  0.31091493  2.78275822  0.00444271  1.04006503  0.89086169
   0.16625734  3.60685931]]


--- 피처별 회귀 계수 ---
lastAction_encoded        3.606859
X                         3.052200
situation_Penalty         2.782758
shot_angle                1.763164
shotType_LeftFoot         1.040065
shotType_RightFoot        0.890862
situation_OpenPlay        0.310915
assist_flag               0.187146
shotType_OtherBodyPart    0.166257
situation_SetPiece        0.004443
home_away                -0.099641
Y                        -0.257908
situation_FromCorner     -0.476730
shot_distance            -2.218133
dtype: float64

--- 절편 (Intercept) ---
-6.989620558048217


## 긍정적 영향 (골 확률 증가 요인) ⚽️

계수가 양수(+)인 피처들은 골 확률을 높이는 요소. 숫자가 클수록 영향력이 더 큼.

- lastAction_encoded (3.6069): 가장 강력한 긍정 요인. 슛 직전에 어떤 행동(예: 드리블, 원투패스 등)을 했는지가 골 확률에 결정적인 영향을 미친다는 것을 모델이 발견. 

- X (3.0522): 경기장 상대 골대에 가까워지는 방향를 의미하는 X값이 클수록 골 확률이 급격히 높아짐. 당연하게도, 상대방 골라인에 가까운 곳에서 찰수록 골이 될 가능성이 높다는 의미.

- situation_Penalty (2.7828): 페널티킥 상황은 골 확률을 매우 크게 높임.

- shot_angle (1.7632): 슈팅 각도가 클수록, 즉 골문을 더 넓게 바라볼수록 골 확률이 높아짐.

- shotType_LeftFoot / RightFoot (1.0401 / 0.8909): 머리나 다른 신체 부위로 슛하는 것보다, 왼발이든 오른발이든 발로 차는 것이 골 확률이 더 높음

## 부정적 영향 (골 확률 감소 요인) 📉

계수가 음수(-)인 피처들은 골 확률을 낮추는 요소. 숫자의 절댓값이 클수록 영향력이 더 큼.

- shot_distance (-2.2181): 가장 강력한 부정 요인. 골문과의 거리가 멀어질수록 골 확률이 급격하게 낮아진다는, 축구의 가장 기본적인 상식을 모델이 정확하게 반영.

- situation_FromCorner (-0.4767): 코너킥으로부터 이어진 슛은 다른 상황(오픈 플레이 등)에 비해 골이 될 확률이 낮음. 코너킥 상황이 혼전 속에서 나오는 질 낮은 슈팅이 많다는 것을 의미.

In [10]:
from sklearn.model_selection import KFold

logit_model = LogisticRegression(random_state=42)

# 교차검증 설정
cv = KFold(n_splits=5, shuffle=True, random_state=42)

mae_scores, r2_scores = [], []

# Fold별 학습 & 평가
for train_idx, test_idx in cv.split(X, y):
    X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]
    xG_test = original_xG.iloc[test_idx]   # <- 비교 대상은 original_xG
    
    logit_model.fit(X_train, y_train)
    
    predicted_xG = logit_model.predict_proba(X_test)[:, 1]
    
    # original_xG와 비교 
    mae = mean_absolute_error(xG_test, predicted_xG)
    r2 = r2_score(xG_test, predicted_xG)
    
    mae_scores.append(mae)
    r2_scores.append(r2)

logit_cv_summary = {
    "MAE_mean": np.mean(mae_scores),
    "R2_mean": np.mean(r2_scores),
}

print("--- Logistic Regression 교차검증 결과 ---")
print(pd.DataFrame([logit_cv_summary]))

--- Logistic Regression 교차검증 결과 ---
   MAE_mean   R2_mean
0  0.056722  0.645342


비슷한 성능이 MAE와 R2가 나오는 것을 확인할 수 있음

### 🤔

- 골키퍼에 관한 정보가 없어 아쉬움
- 골키퍼 또한 골에 큰 영향을 미치기에 슛 당시 골키퍼의 위치에 관한 정보만 있었어도 좀 더 예측이 잘 되었을 것이라 판단